### ISSUE TITLE
Instant loading: MathJax only renders those characters rendered at the last refresh.

### ISSUE THREAD

**Issue by myb6p89r6p** [squidfunk/mkdocs-material#6661]
### Bug description

Say you have two formulas on two pages, formula 1 on page 1, formula 2 on page 2. If you refresh page 1, formula 1 will render properly, but formula 2 will only display those characters that appear in formula 1. Similarly, if you refresh page 2, formula 2 will render properly, but formula 1 will only display those characters that appear in formula 2.

### Related links

- [Reporting a bug](https://squidfunk.github.io/mkdocs-material/contributing/reporting-a-bug/)
- https://squidfunk.github.io/mkdocs-material/reference/math/#mathjax
- https://squidfunk.github.io/mkdocs-material/setup/extensions/python-markdown-extensions/#arithmatex
- https://squidfunk.github.io/mkdocs-material/setup/setting-up-navigation/#instant-loading
- https://github.com/squidfunk/mkdocs-material/issues/2530

### Reproduction

[replication.zip](https://github.com/squidfunk/mkdocs-material/files/13974603/replication.zip)

### Steps to reproduce

1. Go to preview. The math expression in index.md should render fine.
![img-0](https://github.com/squidfunk/mkdocs-material/assets/157003308/6c5a39b7-15c3-4ae3-b080-cec73a7b21fd)
2. Click on Outdex (outdex.md). The math expression in outdex.md will render only partially:
![img-1](https://github.com/squidfunk/mkdocs-material/assets/157003308/29a78529-fe5c-4fcf-85f3-de6feb473f19)
3. Refresh the page. The math expression in outdex.md should now be rendered fully:
![img-2](https://github.com/squidfunk/mkdocs-material/assets/157003308/69ee3d48-5db5-49a8-93fd-97371b1a541f)
4. Return to Home (index.md). Now the math expression in index.md is rendered partially:
![img-3](https://github.com/squidfunk/mkdocs-material/assets/157003308/faa811b9-36f5-4010-bc75-ec76bfbae546)

### Before submitting

- [X] I have read and followed the [bug reporting guidelines](https://squidfunk.github.io/mkdocs-material/contributing/reporting-a-bug/).
- [X] I have attached links to [the documentation](https://squidfunk.github.io/mkdocs-material/), and possibly related [issues](https://github.com/squidfunk/mkdocs-material/issues) and [discussions](https://github.com/squidfunk/mkdocs-material/discussions).
- [X] I assure that I have [removed all customizations](https://squidfunk.github.io/mkdocs-material/contributing/reporting-a-bug/#remove-customizations) before submitting this bug report.
- [X] I have attached a __.zip file__ with a [minimal reproduction](https://squidfunk.github.io/mkdocs-material/guides/creating-a-reproduction/).

**Comment by squidfunk** [comment]
Thanks for reporting. Indeed, `MathJax.typesetPromise()` doesn't appear to be correctly working. If you comment out the following lines from your extra JavaScript:

``` ts
// document$.subscribe(() => { 
//   MathJax.typesetPromise()
// })
```

... and call `MathJax.typesetPromise()` manually in the console, the same thing happens:

<img width="482" alt="Bildschirm­foto 2024-01-18 um 17 15 23" src="https://github.com/squidfunk/mkdocs-material/assets/932156/4d7e9ae9-4634-44a5-8c16-2b748140ef39">

Thus, it is not related to instant navigation, but seems to be caused by MathJax itself.

**Comment by facelessuser** [comment]
@squidfunk I  have never switched over to the official Material MathJax support (never got around to it, and if it ain't broke, don't fix it 🙃 ), but I was having issues until I updated the below. Numbers were not getting reset, I was getting blank blocks, etc.

```js
    MathJax.startup.output.clearCache()
    MathJax.typesetClear()
    MathJax.texReset()
    MathJax.typesetPromise()
```

Just an FYI if it helps

**Comment by facelessuser** [comment]
@squidfunk it may be worth reopening this issue.

**Comment by squidfunk** [comment]
@facelessuser I'm not sure what we should do? This needs to be fixed upstream. We could recommend using the sequence of instructions you suggested in https://github.com/squidfunk/mkdocs-material/issues/6661#issuecomment-1898496201, but we'd first need to get some feedback from the maintainers of MathJax if that really is the way to go.

**Comment by facelessuser** [comment]
Well, I don't think that what I've suggested is too crazy. Since you are doing live reload, MathJax is storing existing numbering and page reference stuff in cache and in its state machine. When you are live reloading, you are changing everything underneath it. MathJax wasn't exactly designed for this per se. When the page reloads, you have to reset the states or things get weird. That is the thinking I came up with when trying this out, unsurprisingly, it cleared up weird numbering and fixed loading.

When using live reload when a normal load occurs, everything is fine, when you swap pages, you get weird stuff. To me, it doesn't seem like the promise isn't working, it seems the expectation is wrong when the page reloads. MathJax is still in memory, and you have to reset it for the next page.

**Comment by squidfunk** [comment]
Thanks! If somebody could get a confirmation from the authors of MathJax that this is the way to go, we're happy to add a note to our documentation, as I understand this is currently something that you found out through fiddling.

**Comment by facelessuser** [comment]
@squidfunk this isn't really done by fiddling, it was done by reading the docs.

I can tell you that `texReset` is absolutely required. It resets the numbering for a page. If you are using dynamic page loading but keeping old numbering, this is a problem and absolutely should be fixed.

> The TeX input jax allows you to automatically number equations. When modifying a page, this can lead to problems as numbered equations may be removed and added; most commonly, duplicate labels lead to issues.
>
> You can reset equation numbering using the command

`typesetClear` is also needed, read about it in the documentation here: https://docs.mathjax.org/en/latest/web/typeset.html#updating-previously-typeset-content.

> If you modify the page to remove content that contains typeset mathematics, you will need to tell MathJax about that so that it knows the typeset math that you are removing is no longer on the page. You do this by using the MathJax.typesetClear() method.

Lastly, we clear the cache for styles as it is based on the types: https://docs.mathjax.org/en/latest/web/typeset.html#obtaining-the-output-stylesheet

> The CommonHTML output jax CSS can be quite large, so the output jax tries to minimize the stylesheet by including only the styles that are actually needed for the mathematics that has been processed by the output jax. That means you should request the stylesheet only after you have typeset the mathematics itself.

While I admit the docs aren't extremely explicit about things, you can infer it from reading the docs. That is how I came up with what I do and why it works. Feel free to also ask the maintainers of MathJax, but I feel the docs provide all the information needed.

**Comment by squidfunk** [comment]
Thanks for the explanation, @facelessuser – I wasn't aware how much research went into this, and how tedious the API is to use, but I guess our use case is not something MathJax is optimizing for. I've updated the documentation in 14d7874d0.

**Comment by facelessuser** [comment]
No worries. It's good to question things, even when it's me 🙃.
